virtualmachine: Add support for Disk Encryption Set parameter

lm-sig opened the pull request:
Fixes https://github.com/ansible-collections/azure/issues/425

Supersedes https://github.com/ansible-collections/azure/pull/1217

##### SUMMARY
I had a requirement to have virtual machines with disks (both OS and data) that used encryption. The VM module did not support provisioning or attaching any disk type with encryption enabled.

<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
Fixes #425

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Feature Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
azure_vm_virtualmachine

##### ADDITIONAL INFORMATION
A customer may set an Azure policy that requires VMs to have disks with encryption. If you attempt to provision a virtual machine with disks without encryption your request will fail.

This change will support both encrypted or unencrypted disks. Currently it requires you to provide the full Azure resource ID for the data encryption set ID.

lm-sig committed:
virtualmachine: Add support for Disk Encryption Set parameter

Fixes https://github.com/ansible-collections/azure/issues/425

lm-sig committed:
virtualmachine: Add support for Disk Encryption Set parameter

Fixes https://github.com/ansible-collections/azure/issues/425

lm-sig committed:
virtualmachine: Add support for Disk Encryption Set parameter

Fixes https://github.com/ansible-collections/azure/issues/425

lm-sig committed:
Merge branch 'feature/vm-disk-encryption-set-v2' of https://github.com/lm-sig/azure into feature/vm-disk-encryption-set-v2

lm-sig commented:
There are currently (2023-11-01) no merge conflicts and no outstanding review issues. Please review as soon possible. Thanks.

Fred-sun committed:
Add required description to document (#1314)

* Add required description to document

* new change for required_if

* Add missing ']'

Fred-sun commented on plugins/modules/azure_rm_virtualmachine.py:
```diff
@@ -306,6 +315,10 @@
                     - ReadOnly
                     - ReadWrite
+                default: ReadOnly
+            disk_encryption_set_id:
```
Missing type and argument 'disk_encryption_set_id' found in data_disks is listed in DOCUMENTATION.options, but not accepted by the module argument_spec

lm-sig replied:
Thank you. I have addressed this. New commit has been pushed.

Fred-sun commented on plugins/modules/azure_rm_virtualmachine.py:
```diff
@@ -306,6 +315,10 @@
                 choices:
                     - ReadOnly
                     - ReadWrite
+                default: ReadOnly
```
Argument 'caching' in argument_spec found in data_disks defines default as (None) but documentation defines default as ('ReadOnly')

lm-sig replied:
Thank you. This was a mistake. I have addressed this. New commit has been pushed.

Fred-sun reviewed (commented):
Please add test cases for the newly added parameters, thank you!

Fred-sun committed:
Delete unnecessary required sets (#1319)

* Delete unnecessary required sets

* Modify azure_rm_virtualnetworkgaway.py

* restore azure_rm_mysqlfirewallrule.py change

Fred-sun committed:
Disable azure_rm_iothub.py test (#1320)

Fred-sun committed:
Bump version to v1.19.0 (#1317)

* Bump version to v1.19.0

* The storage_profile is not a required parameter for creation

* update doc

* update doc

* restore azure_rm_galleryimageversion.py changes

* update doc

---------

Co-authored-by: xuzhang3 <xuzhang3@microsoft.com>

lm-sig committed:
virtualmachine: Pull request fixes for documentation

https://github.com/ansible-collections/azure/pull/1316

lm-sig committed:
virtualmachine: Pull request fixes for documentation

https://github.com/ansible-collections/azure/pull/1316

Fred-sun committed:
Ansible "is match" does not need a ^ (#1321)

ephracis committed:
Add auth_mode to azure_rm_storageblob (#1315)

* Add auth_mode to azure_rm_storageblob

Add the parameter `auth_mode` to the
`azure_rm_storageblob` module, making it possible to create blobs
without using access keys.

* Fix sanity tests

* Add integration tests

Test using auth_mode='login' for storage blobs.
This has been tested on a machine with MSI.

* Remove spaces around equal sign

Fred-sun committed:
Fix caching choices (#1324)

ephracis committed:
Add encryption to azure_rm_galleryimageversion (#1311)

* Add encryption to azure_rm_galleryimageversion

Add the parameter `encryption` to `target_regions` of the
`azure_rm_galleryimageversion` module, making it possible to create
encrypted images.

Fixes #1290

* Fix documentation

* Add disk encryption to gallery image version tests

Add the usage of disk encryption to all regions in the tests of the
module azure_rm_galleryimageversion.

This required a fix, moving the deprecated managed_image to
storage_profile.source_image instead.

ephracis committed:
Add features to azure_rm_galleryimage (#1310)

* Add features to azure_rm_galleryimage

Add the parameter `features` to the
`azure_rm_galleryimage` module, making it possible to create VMs with
Trusted Launch from these images.

* Add integration tests

Switch gallery image definition to Hyper-V
generation V2 and enable TrustedLaunch.

xuzhang3 committed:
Add ansible-core v2.16 (#1305)

* add ansible-core v2.16

* add v2.16

* ignore for v2.16

* ignore for v2.16

* ignore for v2.16

xuzhang3 committed:
bump APIM api version to 2022-08-01 (#1327)

Fred-sun committed:
add hasattr method to verify return value (#1307)

xuzhang3 committed:
Migrate ADGraph to MSGraph (#1325)

* migrate to msgraph

* migrate to msgraph

* migrate to msgraph

* migrate to ms graph API

* ad group info

* app info and group info

* migrate to msgraph

* spn msgraph

* migrate AD to MS graph

* Migrate AAD API to MSGraph API for ServicePrincipal, User and Password (#1280)

* migrate sp and fix application

* migrate azure_rm_adserviceprincipal.py

* migrate azure_rm_aduser_info.py

* migrate azure_rm_adpassword_info.py

* migrate azure_rm_aduser.py

* Migrate azure_rm_adpassword.py and fix azure_rm_adpassword_info.py

* update azure_rm_adserviceprincipal.py

* update azure_rm_adserviceprincipal_info.py

* update azure_rm_aduser.py

* update azure_rm_aduser_info.py

* update azure_rm_adpassword.py

* update azure_rm_adpassword_info.py

* update azure_rm_aduser.py

* update azure_rm_adpassword.py

* update azure_rm_adpassword_info.py

* update azure_rm_adpassword.py to use add_password and remove_password

* update azure_rm_aduser_info.py to add select

* update azure_rm_aduser.py to add select

* update code

* update code

* update code

* update code

* update code

* update code

---------

Co-authored-by: neil-yechenwei <v-cheye@microsoft.com>

* cli auth

* auth with CLI

* Get the existing user after create (#1286)

Co-authored-by: neil-yechenwei <v-cheye@microsoft.com>

* Add advanced query filter to get the existing user by list (#1287)

Co-authored-by: neil-yechenwei <v-cheye@microsoft.com>

* get app by app id

* fix lint

* fix sanity

* fix sanity

* fix sanity

* fix sanity

* deprecate tenant ID

* Mark the property for deprecating and renaming property (#1294)

* Mark the property for deprecating and renaming property

* update code

* update code

---------

Co-authored-by: neil-yechenwei <v-cheye@microsoft.com>

* deprecate tenant ID

* msgraph sdk v1.0.0

* deprecate tenant ID

* update doc

* fix lint

* fix lint

* remove blank space

* fix lint

* fix lint

* fix lint

* remove tenant

* extend test timeout

* update adapp response

---------

Co-authored-by: Neil Ye <yechenwei2007@hotmail.com>
Co-authored-by: neil-yechenwei <v-cheye@microsoft.com>

lm-sig committed:
tests: Add azure_rm_virtualmachine DES case

lm-sig committed:
tests: Add azure_rm_virtualmachine DES case

lm-sig committed:
virtualmachine: Pull request fixes for documentation

https://github.com/ansible-collections/azure/pull/1316

# Conflicts:
#	plugins/modules/azure_rm_virtualmachine.py

lm-sig commented:
I added a test case and updated to the latest 'dev' branch changes. As of today (2023-11-13) there should be no outstanding issues. Thanks.

Fred-sun commented:
@lm-sig Thanks for your update! The 'os_disk_encryption_set_id' was added in #1306 and 'disk_encrytpion_set_id' was added in PR #1309. Thank you!

lm-sig commented:
It's unfortunate Red Hat / @ephracis didn't see this was being worked on for several months, but thank you for the update. I guess this can be closed. Is there an ETA on the 1.19.0 release?
